From 6e8eaa7db03cff96a168d1b78a705c3525b8cafc Mon Sep 17 00:00:00 2001
From: Andre McCurdy <armccurdy@gmail.com>
Date: Thu, 12 Nov 2015 15:40:00 -0800
Subject: [PATCH] nexus/build/os/linuxkernel/os_flags.inc: support separate
 kernel source and build dirs

Also use the same order for kernel header include paths as KBuild does.

Signed-off-by: Andre McCurdy <armccurdy@gmail.com>
---
 nexus/build/os/linuxkernel/os_flags.inc | 31 +++++++++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/nexus/build/os/linuxkernel/os_flags.inc b/nexus/build/os/linuxkernel/os_flags.inc
index 4f53983..908e0e5 100644
--- a/nexus/build/os/linuxkernel/os_flags.inc
+++ b/nexus/build/os/linuxkernel/os_flags.inc
@@ -82,7 +82,11 @@ LINUX_INC ?= $(LINUX)/include
 
 ifeq ($(LINUX_VER_GE_2_6_31),y)
 ifeq ($(LINUX_VER_GE_2_6_37),y)
+ifneq ($(STAGING_KERNEL_BUILDDIR),)
+LINUX_CONFIGURED := $(shell test -f $(STAGING_KERNEL_BUILDDIR)/include/generated/autoconf.h && echo "y")
+else
 LINUX_CONFIGURED := $(shell test -f $(LINUX_INC)/generated/autoconf.h && echo "y")
+endif
 else
 LINUX_CONFIGURED := $(shell test -f $(LINUX_INC)/linux/autoconf.h && echo "y")
 endif
@@ -117,11 +121,23 @@ NEXUS_CFLAGS += -msoft-float -pipe
 NEXUS_CFLAGS += -DLINUX -D__KERNEL__
 ifeq ($(LINUX_VER_GE_2_6_31),y)
 LINUX_ASM_INC ?= $(LINUX)/arch/mips/include/asm/mach-brcmstb
+
+ifneq ($(STAGING_KERNEL_BUILDDIR),)
+NEXUS_CFLAGS += \
+        -I$(LINUX)/arch/mips/include \
+        -I$(STAGING_KERNEL_BUILDDIR)/arch/mips/include/generated \
+        -I$(STAGING_KERNEL_BUILDDIR)/include \
+        -I$(LINUX)/include \
+        -I$(LINUX_ASM_INC) \
+        -I$(LINUX)/arch/mips/include/asm/mach-generic
+else
 NEXUS_CFLAGS += \
     -I$(LINUX_INC) \
     -I$(LINUX)/arch/mips/include \
     -I$(LINUX_ASM_INC) \
     -I$(LINUX)/arch/mips/include/asm/mach-generic
+endif
+
 else
 ifeq ($(filter ${B_REFSW_ARCH}, mips-linux mips-uclibc mips-linux-uclibc), ${B_REFSW_ARCH})
 B_REFSW_TOOLCHAIN_ARCH=mips-linux-uclibc
@@ -156,6 +172,20 @@ endif
 else
 # ARM
 NEXUS_CFLAGS+=-DLINUX -DMODULE -mlong-calls
+
+ifneq ($(STAGING_KERNEL_BUILDDIR),)
+NEXUS_KERNEL_CFLAGS += \
+        -D__KERNEL__ \
+        -I$(LINUX)/arch/arm/include \
+        -I$(STAGING_KERNEL_BUILDDIR)/arch/arm/include/generated \
+        -I$(LINUX)/include \
+        -I$(STAGING_KERNEL_BUILDDIR)/include \
+        -I$(LINUX)/arch/arm/include/uapi \
+        -I$(STAGING_KERNEL_BUILDDIR)/arch/arm/include/generated/uapi \
+        -I$(LINUX)/include/uapi \
+        -I$(STAGING_KERNEL_BUILDDIR)/include/generated/uapi \
+        -include linux/kconfig.h
+else
 NEXUS_KERNEL_CFLAGS += \
         -D__KERNEL__ \
         -I$(LINUX)/include\
@@ -166,6 +196,7 @@ NEXUS_KERNEL_CFLAGS += \
         -I$(LINUX)/arch/arm/include/generated/uapi\
         -I$(LINUX)/include/uapi\
         -I$(LINUX)/include/generated/uapi
+endif
 
 # When building with NEXUS_MODE = "proxy", it seems that some compiler command
 # lines include both NEXUS_CFLAGS and NEXUS_KERNEL_CFLAGS, but some include
-- 
1.9.1

